via/views/api/youtube.py

```diff
@@ -1,1328 +1,42 @@
-from pyramid.httpexceptions import HTTPNotFound
+import logging
+
+from h_pyramid_sentry import report_exception as report_exception_to_sentry
 from pyramid.view import view_config
 
+from via.services import YouTubeService
+
+logger = logging.getLogger(__name__)
+
 
 @view_config(route_name="api.youtube.transcript", permission="api", renderer="json")
 def transcript(request):
     """Return the transcript of a given YouTube video."""
 
     video_id = request.matchdict["video_id"]
 
-    if video_id != "1":
-        raise HTTPNotFound()
+    try:
+        transcript_ = request.find_service(YouTubeService).get_transcript(video_id)
+    except Exception as exc:  # pylint: disable=broad-exception-caught
+        report_exception_to_sentry(exc)
+        logger.exception(exc)
+
+        request.response.status_int = 500
+
+        return {
+            "errors": [
+                {
+                    "status": request.response.status_int,
+                    "code": "failed_to_get_transcript",
+                    "title": "Failed to get transcript from YouTube",
+                    "detail": str(exc),
```

Comment:
We don't know anything about what types of exception object the youtube-transcript-api might raise, so just stringify the whole exception into the error response, leaning on the side of more information rather than less.

Note that I don't expect the frontend to show the `detail` string to the user except perhaps hidden in a "Technical Details" box, mostly the frontend is going to either have to use the generic `title` from our API response or its own hard-coded generic error message.

Reply:
I think we can do a bit more than this if we choose to. All the structured errors this library sends are children of `CouldNotRetrieveTranscript` which have some predicable features.

We can do something like:

```python
    ...
    except CouldNotRetrieveTranscript as err:
        return {
            "errors": [
                {
                    "code": err.__class__.__name__,
                    "title": err.cause,
                    "detail": str(err).strip()
                }
            ]
        }
```

This results in a message like:

```json
{
    "errors": [
        {
            "code": "TranscriptsDisabled",
            "title": "Subtitles are disabled for this video",
            "detail": "Could not retrieve a transcript ... already describe your problem!"
        }
    ]
}
```

The full detail string in this case was:

> Could not retrieve a transcript for the video https://www.youtube.com/watch?v=klsdnjklhjsdfjklsdnfjklsdjklsdnf! This is most likely caused by:\n\nSubtitles are disabled for this video\n\nIf you are sure that the described cause is not responsible for this error and that a transcript should be retrievable, please create an issue at https://github.com/jdepoix/youtube-transcript-api/issues. Please add which version of youtube_transcript_api you are using and provide the information needed to replicate the error. Also make sure that there are no open issues which already describe your problem!

Reply:
:+1: I think this is worth doing just to provide slightly more informative / dynamic error responses to the frontend: https://github.com/hypothesis/via/pull/1010/commits/d91fd95397a388cbff4cca6257a0f6fe61a5e45b

With my two test cases from the PR description this now produces:

```terminal
$ http GET 'http://localhost:9083/api/youtube/transcript/foo' Authorization:"Bearer eyJ***V-g"
HTTP/1.1 500 Internal Server Error
Connection: keep-alive
Content-Length: 691
Content-Type: application/json
Date: Wed, 28 Jun 2023 11:39:54 GMT
Server: nginx/1.17.6
X-Abuse-Policy: https://web.hypothes.is/abuse-policy/
X-Complaints-To: https://web.hypothes.is/report-abuse/
X-Robots-Tag: noindex, nofollow

{
    "errors": [
        {
            "code": "TranscriptsDisabled",
            "detail": "Could not retrieve a transcript for the video https://www.youtube.com/watch?v=foo! This is most likely caused by:\n\nSubtitles are disabled for this video\n\nIf you are sure that the described cause is not responsible for this error and that a transcript should be retrievable, please create an issue at https://github.com/jdepoix/youtube-transcript-api/issues. Please add which version of youtube_transcript_api you are using and provide the information needed to replicate the error. Also make sure that there are no open issues which already describe your problem!",
            "status": 500,
            "title": "Subtitles are disabled for this video"
        }
    ]
}

$ http GET 'http://localhost:9083/api/youtube/transcript/TbzZIMQC6vk' Authorization:"Bearer eyJ***V-g"
HTTP/1.1 500 Internal Server Error
Connection: keep-alive
Content-Length: 409
Content-Type: application/json
Date: Wed, 28 Jun 2023 11:40:43 GMT
Server: nginx/1.17.6
X-Abuse-Policy: https://web.hypothes.is/abuse-policy/
X-Complaints-To: https://web.hypothes.is/report-abuse/
X-Robots-Tag: noindex, nofollow

{
    "errors": [
        {
            "code": "ConnectionError",
            "detail": "HTTPSConnectionPool(host='www.youtube.com', port=443): Max retries exceeded with url: /watch?v=TbzZIMQC6vk (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7f3a043e1eb0>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution'))",
            "status": 500,
            "title": "Failed to get transcript from YouTube"
        }
    ]
}
```

Note that the `"status": 500` in the error body is required by the JSON:API spec